test/server/http/admin_test.cc

```diff
@@ -293,5 +294,121 @@ TEST(PrometheusStatsFormatter, FormattedTags) {
   EXPECT_EQ(expected, actual);
 }
 
+TEST(PrometheusStatsFormatter, MetricNameCollison) {
+
+  // Create two counters and two gauges with each pair having the same name,
+  // but having different tag names and values.
+  //`statsAsPrometheus()` should return two implying it found two unique stat names
+
+  Stats::HeapRawStatDataAllocator alloc;
+  std::list<Stats::CounterSharedPtr> counters;
+  std::list<Stats::GaugeSharedPtr> gauges;
+
+  {
+    std::string name = "cluster.test_cluster_1.upstream_cx_total";
+    std::vector<Stats::Tag> cluster_tags;
+    Stats::Tag tag = {"a.tag-name", "a.tag-value"};
+    cluster_tags.push_back(tag);
+    Stats::RawStatData* data = alloc.alloc(name);
+    Stats::CounterSharedPtr c(
+        new Stats::CounterImpl(*data, alloc, std::string(name), std::move(cluster_tags)));
+    counters.push_back(c);
+  }
+
+  {
+    std::string name = "cluster.test_cluster_1.upstream_cx_total";
+    std::vector<Stats::Tag> cluster_tags;
+    Stats::Tag tag = {"another_tag_name", "another_tag-value"};
+    cluster_tags.push_back(tag);
+    Stats::RawStatData* data = alloc.alloc(name);
+    Stats::CounterSharedPtr c(
+        new Stats::CounterImpl(*data, alloc, std::string(name), std::move(cluster_tags)));
+    counters.push_back(c);
```

Comment:
This scenario seems contrived. How does it happen in practice?

Reply:
counterImpl/GuageImpl are derived classes from counter/gauge class so i create two counters and two gauges and pass them with same or different metric names in two different tests and check if metric name was same then gauge/counter are inserted only once for each unique metric name . Promlint tool was failing due to same metric name printed multiple times.

Reply:
I understand how this happens in the test, I'm curious though how this happens in actual Envoy use. How did you discover this? What were the offending counters?

Reply:
The envoy_listener_http_downstream_rq_xx counter appears many times, each time with a different set of tags and metrics that come with that tag (dimension). 

For each upstream server you get 5 envoy_listener_http_downstream_rq_xx counters: one for each http responsecode (1, 2, 3, 4, 5) as the value for the "envoy_response_code_class" tag.